<a href="https://colab.research.google.com/github/youngcheolman/youngcheol-Park/blob/main/%EB%8B%A4%EC%9D%B4%EC%96%B4%ED%8A%B8_%EC%A0%95%EB%A6%AC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

df = pd.read_excel("/content/drive/MyDrive/online_diet.xlsx")
df

seq_num          id  social_id1 zip_code  occupation  sex  height  \
0          28    webinner    740101.0  137-070         6.0  1.0   170.0   
1          29    webinner    740101.0  137-070         6.0  1.0   170.0   
2          30    webinner    740101.0  137-070         6.0  1.0   170.0   
3          31    webinner    740101.0  137-070         6.0  1.0   170.0   
4          32    webinner    740101.0  137-070         6.0  1.0   170.0   
...       ...         ...         ...      ...         ...  ...     ...   
4453     2011  0162655575    830219.0  135-080        15.0  2.0   165.0   
4454     2012      12girl    781007.0  136-043        17.0  2.0   160.0   
4455     2013  onlyonelhm    880330.0  327-030        15.0  2.0   159.0   
4456     2014   morebest1    760709.0  0000000        12.0  1.0   177.0   
4457     2015      kiya80    770426.0  660-322        12.0  2.0   156.0   

      weight  married dise  ...         loss_part diet_purpose     diet   bmi  \
0       62.0      0.0  NaN  ...              1, 2          3.0        4  21.0   
1       62.0      0.0  NaN  ...              1, 2          3.0        4  21.0   
2       62.0      0.0  NaN  ...              1, 2          3.0        4  21.0   
3       62.0      0.0  NaN  ...              1, 2          3.0        4  21.0   
4       62.0      0.0  NaN  ...              1, 2          3.0        4  21.0   
...      ...      ...  ...  ...               ...          ...      ...   ...   
4453    49.0      0.0  NaN  ...              3, 6          1.0      NaN  17.0   
4454    63.0      0.0    2  ...              1, 6          3.0      NaN  24.0   
4455    58.0      0.0  NaN  ...  1, 2, 3, 4, 6, 7          1.0      NaN  22.0   
4456   109.0      0.0  NaN  ...        1, 2, 4, 8          3.0  1, 4, 5  34.0   
4457    49.0      0.0    1  ...              1, 6          2.0        4  20.0   

     order_id        pfid                 p_name  p_price  \
0       16560  MDGLIKE001  오레그 O다리,X다리 교정벨트(고급형)    27000   
1       16926  MDGLIKE001  오레그 O다리,X다리 교정벨트(고급형)    27000   
2       16941  MDGLIKE001  오레그 O다리,X다리 교정벨트(고급형)    27000   
3       16944  MDGLIKE001  오레그 O다리,X다리 교정벨트(고급형)    27000   
4       16967  MDGBEAT004                아로마 장미향    20000   
...       ...         ...                    ...      ...   
4453    21654  MDEHRWE008          파워세라믹종아리 다이어트    29900   
4454    21656  PDFMEAL014             풀무원 드림다이어트   300000   
4455    21664  MDEHRWE008          파워세라믹종아리 다이어트    29900   
4456    21665  MDFMEAL007             대상다이어트프로그램    70000   
4457    21666  PDFMEAL016            깜짝변신 다이어트세트   130000   

                p_created p_quantity  
0     2001-05-17  1:10:38          1  
1     2001-05-17  1:26:33          1  
2     2001-05-17  2:19:23          1  
3     2001-05-17  2:27:24          1  
4     2001-05-17  4:25:11          1  
...                   ...        ...  
4453  2001-07-05  7:17:10          1  
4454  2001-07-05  8:02:18          1  
4455  2001-07-06  1:17:45          1  
4456  2001-07-06  1:50:12          1  
4457  2001-07-06  2:27:39          1  

[4458 rows x 24 columns]

In [2]:
df.drop(['seq_num','zip_code','want_diet','order_id'], axis=1, inplace=True)
# 일련번호, 우편번호,희망체중, 주문번호를 뺀 데이터

In [3]:
df

id  social_id1  occupation  sex  height  weight  married dise  \
0       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
1       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
2       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
3       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
4       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
...          ...         ...         ...  ...     ...     ...      ...  ...   
4453  0162655575    830219.0        15.0  2.0   165.0    49.0      0.0  NaN   
4454      12girl    781007.0        17.0  2.0   160.0    63.0      0.0    2   
4455  onlyonelhm    880330.0        15.0  2.0   159.0    58.0      0.0  NaN   
4456   morebest1    760709.0        12.0  1.0   177.0   109.0      0.0  NaN   
4457      kiya80    770426.0        12.0  2.0   156.0    49.0      0.0    1   

                   created waist heap         loss_part  diet_purpose  \
0      2000-05-12  3:38:49   NaN  NaN              1, 2           3.0   
1      2000-05-12  3:38:49   NaN  NaN              1, 2           3.0   
2      2000-05-12  3:38:49   NaN  NaN              1, 2           3.0   
3      2000-05-12  3:38:49   NaN  NaN              1, 2           3.0   
4      2000-05-12  3:38:49   NaN  NaN              1, 2           3.0   
...                    ...   ...  ...               ...           ...   
4453  2001-03-06  11:42:36   NaN  NaN              3, 6           1.0   
4454   2000-11-01  9:36:21   NaN  NaN              1, 6           3.0   
4455  2001-01-27  10:49:59    29  NaN  1, 2, 3, 4, 6, 7           1.0   
4456  2001-05-01  11:36:13    40   42        1, 2, 4, 8           3.0   
4457   2001-07-06  2:11:54    27   34              1, 6           2.0   

         diet   bmi        pfid                 p_name  p_price  \
0           4  21.0  MDGLIKE001  오레그 O다리,X다리 교정벨트(고급형)    27000   
1           4  21.0  MDGLIKE001  오레그 O다리,X다리 교정벨트(고급형)    27000   
2           4  21.0  MDGLIKE001  오레그 O다리,X다리 교정벨트(고급형)    27000   
3           4  21.0  MDGLIKE001  오레그 O다리,X다리 교정벨트(고급형)    27000   
4           4  21.0  MDGBEAT004                아로마 장미향    20000   
...       ...   ...         ...                    ...      ...   
4453      NaN  17.0  MDEHRWE008          파워세라믹종아리 다이어트    29900   
4454      NaN  24.0  PDFMEAL014             풀무원 드림다이어트   300000   
4455      NaN  22.0  MDEHRWE008          파워세라믹종아리 다이어트    29900   
4456  1, 4, 5  34.0  MDFMEAL007             대상다이어트프로그램    70000   
4457        4  20.0  PDFMEAL016            깜짝변신 다이어트세트   130000   

                p_created  p_quantity  
0     2001-05-17  1:10:38           1  
1     2001-05-17  1:26:33           1  
2     2001-05-17  2:19:23           1  
3     2001-05-17  2:27:24           1  
4     2001-05-17  4:25:11           1  
...                   ...         ...  
4453  2001-07-05  7:17:10           1  
4454  2001-07-05  8:02:18           1  
4455  2001-07-06  1:17:45           1  
4456  2001-07-06  1:50:12           1  
4457  2001-07-06  2:27:39           1  

[4458 rows x 20 columns]

####2002년 현재 시점 기준 
#2001년 가정 데이터

In [4]:
#주민번호 앞자리에서 파생변수 age group을 만들어냄 
df['age_group'] = 0
for i in range(0, len(df)):
    if df.loc[i, 'social_id1'] > 911231 :
        df.loc[i, 'age_group'] = '10'
    elif df.loc[i, 'social_id1'] > 811231 :
        df.loc[i, 'age_group'] = '20'
    elif df.loc[i, 'social_id1'] > 711231 :
        df.loc[i, 'age_group'] = '30'
    elif df.loc[i, 'social_id1'] > 611231 :
        df.loc[i, 'age_group'] = '40'
    elif df.loc[i, 'social_id1'] > 511231 :
        df.loc[i, 'age_group'] = '50'
    else :
         df.loc[i, 'age_group'] = '60'

df

id  social_id1  occupation  sex  height  weight  married dise  \
0       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
1       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
2       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
3       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
4       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
...          ...         ...         ...  ...     ...     ...      ...  ...   
4453  0162655575    830219.0        15.0  2.0   165.0    49.0      0.0  NaN   
4454      12girl    781007.0        17.0  2.0   160.0    63.0      0.0    2   
4455  onlyonelhm    880330.0        15.0  2.0   159.0    58.0      0.0  NaN   
4456   morebest1    760709.0        12.0  1.0   177.0   109.0      0.0  NaN   
4457      kiya80    770426.0        12.0  2.0   156.0    49.0      0.0    1   

                   created waist  ...         loss_part diet_purpose     diet  \
0      2000-05-12  3:38:49   NaN  ...              1, 2          3.0        4   
1      2000-05-12  3:38:49   NaN  ...              1, 2          3.0        4   
2      2000-05-12  3:38:49   NaN  ...              1, 2          3.0        4   
3      2000-05-12  3:38:49   NaN  ...              1, 2          3.0        4   
4      2000-05-12  3:38:49   NaN  ...              1, 2          3.0        4   
...                    ...   ...  ...               ...          ...      ...   
4453  2001-03-06  11:42:36   NaN  ...              3, 6          1.0      NaN   
4454   2000-11-01  9:36:21   NaN  ...              1, 6          3.0      NaN   
4455  2001-01-27  10:49:59    29  ...  1, 2, 3, 4, 6, 7          1.0      NaN   
4456  2001-05-01  11:36:13    40  ...        1, 2, 4, 8          3.0  1, 4, 5   
4457   2001-07-06  2:11:54    27  ...              1, 6          2.0        4   

       bmi        pfid                 p_name p_price            p_created  \
0     21.0  MDGLIKE001  오레그 O다리,X다리 교정벨트(고급형)   27000  2001-05-17  1:10:38   
1     21.0  MDGLIKE001  오레그 O다리,X다리 교정벨트(고급형)   27000  2001-05-17  1:26:33   
2     21.0  MDGLIKE001  오레그 O다리,X다리 교정벨트(고급형)   27000  2001-05-17  2:19:23   
3     21.0  MDGLIKE001  오레그 O다리,X다리 교정벨트(고급형)   27000  2001-05-17  2:27:24   
4     21.0  MDGBEAT004                아로마 장미향   20000  2001-05-17  4:25:11   
...    ...         ...                    ...     ...                  ...   
4453  17.0  MDEHRWE008          파워세라믹종아리 다이어트   29900  2001-07-05  7:17:10   
4454  24.0  PDFMEAL014             풀무원 드림다이어트  300000  2001-07-05  8:02:18   
4455  22.0  MDEHRWE008          파워세라믹종아리 다이어트   29900  2001-07-06  1:17:45   
4456  34.0  MDFMEAL007             대상다이어트프로그램   70000  2001-07-06  1:50:12   
4457  20.0  PDFMEAL016            깜짝변신 다이어트세트  130000  2001-07-06  2:27:39   

     p_quantity  age_group  
0             1         30  
1             1         30  
2             1         30  
3             1         30  
4             1         30  
...         ...        ...  
4453          1         20  
4454          1         30  
4455          1         20  
4456          1         30  
4457          1         30  

[4458 rows x 21 columns]

In [5]:
df['age_group'].value_counts()

30    2994
20     718
40     597
50     122
60      27
Name: age_group, dtype: int64

In [6]:
df['married'].value_counts()

0.0    3688
1.0     768
Name: married, dtype: int64

In [7]:
df['sex'].value_counts()

2.0    4159
1.0     297
Name: sex, dtype: int64

##주요 고객층이 30대 미혼 여성인 것을 파악

In [8]:
#p_created와 created에 대한 파생변수를 계절 변수로 만들건데 이게 문자열 데이터라서 만들기 위해서 datatype을 시계열로 변환

df['p_created'] = pd.to_datetime(df['p_created'])
df['created'] = pd.to_datetime(df['created'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4458 entries, 0 to 4457
Data columns (total 21 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   id            4458 non-null   object        
 1   social_id1    4457 non-null   float64       
 2   occupation    4456 non-null   float64       
 3   sex           4456 non-null   float64       
 4   height        4456 non-null   float64       
 5   weight        4456 non-null   float64       
 6   married       4456 non-null   float64       
 7   dise          2930 non-null   object        
 8   created       4458 non-null   datetime64[ns]
 9   waist         3135 non-null   object        
 10  heap          1899 non-null   object        
 11  loss_part     4206 non-null   object        
 12  diet_purpose  4183 non-null   float64       
 13  diet          3411 non-null   object        
 14  bmi           4456 non-null   float64       
 15  pfid          4458 non-null   object  

In [9]:
df['product buy season'] = df['p_created'].dt.month
df['join season'] = df['created'].dt.month
df

id  social_id1  occupation  sex  height  weight  married dise  \
0       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
1       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
2       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
3       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
4       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
...          ...         ...         ...  ...     ...     ...      ...  ...   
4453  0162655575    830219.0        15.0  2.0   165.0    49.0      0.0  NaN   
4454      12girl    781007.0        17.0  2.0   160.0    63.0      0.0    2   
4455  onlyonelhm    880330.0        15.0  2.0   159.0    58.0      0.0  NaN   
4456   morebest1    760709.0        12.0  1.0   177.0   109.0      0.0  NaN   
4457      kiya80    770426.0        12.0  2.0   156.0    49.0      0.0    1   

                 created waist  ...     diet   bmi        pfid  \
0    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGLIKE001   
1    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGLIKE001   
2    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGLIKE001   
3    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGLIKE001   
4    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGBEAT004   
...                  ...   ...  ...      ...   ...         ...   
4453 2001-03-06 11:42:36   NaN  ...      NaN  17.0  MDEHRWE008   
4454 2000-11-01 09:36:21   NaN  ...      NaN  24.0  PDFMEAL014   
4455 2001-01-27 10:49:59    29  ...      NaN  22.0  MDEHRWE008   
4456 2001-05-01 11:36:13    40  ...  1, 4, 5  34.0  MDFMEAL007   
4457 2001-07-06 02:11:54    27  ...        4  20.0  PDFMEAL016   

                     p_name  p_price           p_created p_quantity  \
0     오레그 O다리,X다리 교정벨트(고급형)    27000 2001-05-17 01:10:38          1   
1     오레그 O다리,X다리 교정벨트(고급형)    27000 2001-05-17 01:26:33          1   
2     오레그 O다리,X다리 교정벨트(고급형)    27000 2001-05-17 02:19:23          1   
3     오레그 O다리,X다리 교정벨트(고급형)    27000 2001-05-17 02:27:24          1   
4                   아로마 장미향    20000 2001-05-17 04:25:11          1   
...                     ...      ...                 ...        ...   
4453          파워세라믹종아리 다이어트    29900 2001-07-05 07:17:10          1   
4454             풀무원 드림다이어트   300000 2001-07-05 08:02:18          1   
4455          파워세라믹종아리 다이어트    29900 2001-07-06 01:17:45          1   
4456             대상다이어트프로그램    70000 2001-07-06 01:50:12          1   
4457            깜짝변신 다이어트세트   130000 2001-07-06 02:27:39          1   

      age_group product buy season  join season  
0            30                  5            5  
1            30                  5            5  
2            30                  5            5  
3            30                  5            5  
4            30                  5            5  
...         ...                ...          ...  
4453         20                  7            3  
4454         30                  7           11  
4455         20                  7            1  
4456         30                  7            5  
4457         30                  7            7  

[4458 rows x 23 columns]

In [10]:
# 1 = 봄, 2 = 여름, 3 = 가을, 4 = 겨울

df['product buy season'] = df['product buy season'].replace(1,"4")
df['product buy season'] = df['product buy season'].replace(2,"4")
df['product buy season'] = df['product buy season'].replace(3,"1")
df['product buy season'] = df['product buy season'].replace(4,"1")
df['product buy season'] = df['product buy season'].replace(5,"1")
df['product buy season'] = df['product buy season'].replace(6,"2")
df['product buy season'] = df['product buy season'].replace(7,"2")
df['product buy season'] = df['product buy season'].replace(8,"2")
df['product buy season'] = df['product buy season'].replace(9,"3")
df['product buy season'] = df['product buy season'].replace(10,"3")
df['product buy season'] = df['product buy season'].replace(11,"3")
df['product buy season'] = df['product buy season'].replace(12,"4")
df['join season'] = df['join season'].replace(1,"4")
df['join season'] = df['join season'].replace(2,"4")
df['join season'] = df['join season'].replace(3,"1")
df['join season'] = df['join season'].replace(4,"1")
df['join season'] = df['join season'].replace(5,"1")
df['join season'] = df['join season'].replace(6,"2")
df['join season'] = df['join season'].replace(7,"2")
df['join season'] = df['join season'].replace(8,"2")
df['join season'] = df['join season'].replace(9,"3")
df['join season'] = df['join season'].replace(10,"3")
df['join season'] = df['join season'].replace(11,"3")
df['join season'] = df['join season'].replace(12,"4")
df

#for i in range(1,13):
#  if i <= 2 or i ==12:
#    df['join season'] = df['join season'].replace(i,"4")
#    df['product buy season'] = df['product buy season'].replace(1,"4")
#  elif 2 < i < 6 :
#    df['join season'] = df['join season'].replace(i,"1")
#    df['product buy season'] = df['product buy season'].replace(3,"1")
#  elif 5 < i < 9  :
#    df['join season'] = df['join season'].replace(i,"2")
#    df['product buy season'] = df['product buy season'].replace(6,"2")
#  elif 8 < i <12 :
#    df['product buy season'] = df['product buy season'].replace(10,"3")
#    df['join season'] = df['join season'].replace(i,"4")

id  social_id1  occupation  sex  height  weight  married dise  \
0       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
1       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
2       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
3       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
4       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
...          ...         ...         ...  ...     ...     ...      ...  ...   
4453  0162655575    830219.0        15.0  2.0   165.0    49.0      0.0  NaN   
4454      12girl    781007.0        17.0  2.0   160.0    63.0      0.0    2   
4455  onlyonelhm    880330.0        15.0  2.0   159.0    58.0      0.0  NaN   
4456   morebest1    760709.0        12.0  1.0   177.0   109.0      0.0  NaN   
4457      kiya80    770426.0        12.0  2.0   156.0    49.0      0.0    1   

                 created waist  ...     diet   bmi        pfid  \
0    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGLIKE001   
1    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGLIKE001   
2    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGLIKE001   
3    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGLIKE001   
4    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGBEAT004   
...                  ...   ...  ...      ...   ...         ...   
4453 2001-03-06 11:42:36   NaN  ...      NaN  17.0  MDEHRWE008   
4454 2000-11-01 09:36:21   NaN  ...      NaN  24.0  PDFMEAL014   
4455 2001-01-27 10:49:59    29  ...      NaN  22.0  MDEHRWE008   
4456 2001-05-01 11:36:13    40  ...  1, 4, 5  34.0  MDFMEAL007   
4457 2001-07-06 02:11:54    27  ...        4  20.0  PDFMEAL016   

                     p_name  p_price           p_created p_quantity  \
0     오레그 O다리,X다리 교정벨트(고급형)    27000 2001-05-17 01:10:38          1   
1     오레그 O다리,X다리 교정벨트(고급형)    27000 2001-05-17 01:26:33          1   
2     오레그 O다리,X다리 교정벨트(고급형)    27000 2001-05-17 02:19:23          1   
3     오레그 O다리,X다리 교정벨트(고급형)    27000 2001-05-17 02:27:24          1   
4                   아로마 장미향    20000 2001-05-17 04:25:11          1   
...                     ...      ...                 ...        ...   
4453          파워세라믹종아리 다이어트    29900 2001-07-05 07:17:10          1   
4454             풀무원 드림다이어트   300000 2001-07-05 08:02:18          1   
4455          파워세라믹종아리 다이어트    29900 2001-07-06 01:17:45          1   
4456             대상다이어트프로그램    70000 2001-07-06 01:50:12          1   
4457            깜짝변신 다이어트세트   130000 2001-07-06 02:27:39          1   

      age_group product buy season  join season  
0            30                  1            1  
1            30                  1            1  
2            30                  1            1  
3            30                  1            1  
4            30                  1            1  
...         ...                ...          ...  
4453         20                  2            1  
4454         30                  2            3  
4455         20                  2            4  
4456         30                  2            1  
4457         30                  2            2  

[4458 rows x 23 columns]

In [11]:
df.reset_index(drop=True, inplace=True) 
df

id  social_id1  occupation  sex  height  weight  married dise  \
0       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
1       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
2       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
3       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
4       webinner    740101.0         6.0  1.0   170.0    62.0      0.0  NaN   
...          ...         ...         ...  ...     ...     ...      ...  ...   
4453  0162655575    830219.0        15.0  2.0   165.0    49.0      0.0  NaN   
4454      12girl    781007.0        17.0  2.0   160.0    63.0      0.0    2   
4455  onlyonelhm    880330.0        15.0  2.0   159.0    58.0      0.0  NaN   
4456   morebest1    760709.0        12.0  1.0   177.0   109.0      0.0  NaN   
4457      kiya80    770426.0        12.0  2.0   156.0    49.0      0.0    1   

                 created waist  ...     diet   bmi        pfid  \
0    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGLIKE001   
1    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGLIKE001   
2    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGLIKE001   
3    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGLIKE001   
4    2000-05-12 03:38:49   NaN  ...        4  21.0  MDGBEAT004   
...                  ...   ...  ...      ...   ...         ...   
4453 2001-03-06 11:42:36   NaN  ...      NaN  17.0  MDEHRWE008   
4454 2000-11-01 09:36:21   NaN  ...      NaN  24.0  PDFMEAL014   
4455 2001-01-27 10:49:59    29  ...      NaN  22.0  MDEHRWE008   
4456 2001-05-01 11:36:13    40  ...  1, 4, 5  34.0  MDFMEAL007   
4457 2001-07-06 02:11:54    27  ...        4  20.0  PDFMEAL016   

                     p_name  p_price           p_created p_quantity  \
0     오레그 O다리,X다리 교정벨트(고급형)    27000 2001-05-17 01:10:38          1   
1     오레그 O다리,X다리 교정벨트(고급형)    27000 2001-05-17 01:26:33          1   
2     오레그 O다리,X다리 교정벨트(고급형)    27000 2001-05-17 02:19:23          1   
3     오레그 O다리,X다리 교정벨트(고급형)    27000 2001-05-17 02:27:24          1   
4                   아로마 장미향    20000 2001-05-17 04:25:11          1   
...                     ...      ...                 ...        ...   
4453          파워세라믹종아리 다이어트    29900 2001-07-05 07:17:10          1   
4454             풀무원 드림다이어트   300000 2001-07-05 08:02:18          1   
4455          파워세라믹종아리 다이어트    29900 2001-07-06 01:17:45          1   
4456             대상다이어트프로그램    70000 2001-07-06 01:50:12          1   
4457            깜짝변신 다이어트세트   130000 2001-07-06 02:27:39          1   

      age_group product buy season  join season  
0            30                  1            1  
1            30                  1            1  
2            30                  1            1  
3            30                  1            1  
4            30                  1            1  
...         ...                ...          ...  
4453         20                  2            1  
4454         30                  2            3  
4455         20                  2            4  
4456         30                  2            1  
4457         30                  2            2  

[4458 rows x 23 columns]

In [12]:
#product buy season  1 = 봄, 2 = 여름, 3 = 가을, 4 = 겨울

df['product buy season'].value_counts()

2    4138
1     320
Name: product buy season, dtype: int64

In [13]:
df['join season'].value_counts()

2    2429
3     826
1     822
4     381
Name: join season, dtype: int64

In [84]:
df['occupation'].value_counts()

15    1797
17     603
12     544
3      495
14     406
1      178
6      130
7      106
16      74
4       64
11      18
13      16
8       12
5        4
10       4
9        3
2        1
Name: occupation, dtype: int64

사람들이 제품을 구매하는 시기는 여름이 4138, 봄이 320

그리고 우리 회사에 가입하는 시기는 여름 2429, 가을이 826이다


##이를 통해 사람들이 여름에 갑자기 다이어트를 시작하려고 하는 것을 알 수 있었다.

갑자기 하는 다이어트는 건강에 좋지도 않고, 요요의 위험성도 있다 
#그래서 우리는 주제를 장기적이고 건강한 다이어트 프로젝트를 구상해보았다.

In [14]:
df1 = df[(df.sex == 2.0) & (df.age_group == '30') & (df.married == 0)]
df2 = df[(df.sex != 2.0) | (df.age_group != '30') | (df.married != 0)]
df1["label"] = 1
df2["label"] = 0

<ipython-input-14-4cf6a8a63c75>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1["label"] = 1
<ipython-input-14-4cf6a8a63c75>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2["label"] = 0


In [15]:
df3 = pd.concat([df1, df2], axis = 0)
df = df3

df

id  social_id1  occupation  sex  height  weight  married dise  \
6         bluezoo    761220.0        17.0  2.0   167.0    58.0      0.0    2   
7     sunsook1120    751120.0        17.0  2.0   168.0    62.0      0.0    3   
11        minimee    740531.0        15.0  2.0   160.0    46.0      0.0  NaN   
12        hammy99    800808.0        15.0  2.0   164.0    51.0      0.0  NaN   
14           ippn    770415.0        17.0  2.0   160.0    47.0      0.0    2   
...           ...         ...         ...  ...     ...     ...      ...  ...   
4447         ekoo    770830.0        15.0  1.0   178.0    70.0      0.0  NaN   
4451      hjlee90    711023.0         7.0  2.0   172.0    64.0      1.0  NaN   
4453   0162655575    830219.0        15.0  2.0   165.0    49.0      0.0  NaN   
4455   onlyonelhm    880330.0        15.0  2.0   159.0    58.0      0.0  NaN   
4456    morebest1    760709.0        12.0  1.0   177.0   109.0      0.0  NaN   

                 created waist  ...   bmi        pfid         p_name p_price  \
6    2000-08-18 01:59:10    28  ...  21.0  MDEPART007  파워세라믹종아리 다이어트   39000   
7    2000-10-20 11:32:54   NaN  ...  21.0  MDGLIKE013   뷰티라인(땡김이)-멀티   22000   
11   2000-06-14 11:16:57    26  ...  17.0  PDFFUNC001    내추럴하우스 라인-업   60000   
12   2000-12-01 07:56:07   NaN  ...  19.0  MDGLIKE006          매직마스크   24900   
14   2000-10-27 05:31:32    27  ...  18.0  MDGLIKE012   뷰티라인(땡김이)-심플   18000   
...                  ...   ...  ...   ...         ...            ...     ...   
4447 2001-07-05 02:30:13    28  ...  22.0  MDEETCC015         스폐셜킥보드   35900   
4451 2001-07-05 03:02:21    29  ...  21.0  PDFFUNC007    깜짝변신 다이어트세트  130000   
4453 2001-03-06 11:42:36   NaN  ...  17.0  MDEHRWE008  파워세라믹종아리 다이어트   29900   
4455 2001-01-27 10:49:59    29  ...  22.0  MDEHRWE008  파워세라믹종아리 다이어트   29900   
4456 2001-05-01 11:36:13    40  ...  34.0  MDFMEAL007     대상다이어트프로그램   70000   

               p_created p_quantity age_group  product buy season join season  \
6    2001-05-28 08:26:02          1        30                   1           2   
7    2001-05-28 10:12:39          1        30                   1           3   
11   2001-05-28 09:10:42          1        30                   1           2   
12   2001-05-27 01:36:40          1        30                   1           4   
14   2001-05-29 01:28:42          1        30                   1           3   
...                  ...        ...       ...                 ...         ...   
4447 2001-07-05 02:36:53          1        30                   2           2   
4451 2001-07-05 03:09:10          1        40                   2           2   
4453 2001-07-05 07:17:10          1        20                   2           1   
4455 2001-07-06 01:17:45          1        20                   2           4   
4456 2001-07-06 01:50:12          1        30                   2           1   

      label  
6         1  
7         1  
11        1  
12        1  
14        1  
...     ...  
4447      0  
4451      0  
4453      0  
4455      0  
4456      0  

[4458 rows x 24 columns]

In [16]:
df.isnull().sum()

id                       0
social_id1               1
occupation               2
sex                      2
height                   2
weight                   2
married                  2
dise                  1528
created                  0
waist                 1323
heap                  2559
loss_part              252
diet_purpose           275
diet                  1047
bmi                      2
pfid                     0
p_name                   0
p_price                  0
p_created                0
p_quantity               0
age_group                0
product buy season       0
join season              0
label                    0
dtype: int64

## 결측치 전처리


waist와 heap은 bmi로 평균적인 추정이 가능하기 때문에 2개의 항목은 삭제 

loss_part 원핫 인코딩처럼 1~9까지 항목을 더 만들어 분류

diet_purpose는 대부분의 사람들이 원하는 부위가 비슷하다는 판단에 평균으로 대체

diet와 dise는 사람들의 개인적인 경험과 질병을 표시한 데이터기 때문에 결측값들은 그냥 0으로 표기 하여 나타내려고 함.

In [17]:
df.drop(['waist','heap'], axis=1, inplace=True)

In [18]:
#diet_purpose의 결측값을 최빈값 1로 대체
df["diet_purpose"].fillna("1", inplace = True)
df["dise"].fillna("0", inplace = True)
df["loss_part"].fillna("0", inplace = True)

In [19]:
df

id  social_id1  occupation  sex  height  weight  married dise  \
6         bluezoo    761220.0        17.0  2.0   167.0    58.0      0.0    2   
7     sunsook1120    751120.0        17.0  2.0   168.0    62.0      0.0    3   
11        minimee    740531.0        15.0  2.0   160.0    46.0      0.0    0   
12        hammy99    800808.0        15.0  2.0   164.0    51.0      0.0    0   
14           ippn    770415.0        17.0  2.0   160.0    47.0      0.0    2   
...           ...         ...         ...  ...     ...     ...      ...  ...   
4447         ekoo    770830.0        15.0  1.0   178.0    70.0      0.0    0   
4451      hjlee90    711023.0         7.0  2.0   172.0    64.0      1.0    0   
4453   0162655575    830219.0        15.0  2.0   165.0    49.0      0.0    0   
4455   onlyonelhm    880330.0        15.0  2.0   159.0    58.0      0.0    0   
4456    morebest1    760709.0        12.0  1.0   177.0   109.0      0.0    0   

                 created            loss_part  ...   bmi        pfid  \
6    2000-08-18 01:59:10           4, 5, 6, 8  ...  21.0  MDEPART007   
7    2000-10-20 11:32:54  1, 2, 3, 4, 5, 6, 7  ...  21.0  MDGLIKE013   
11   2000-06-14 11:16:57              1, 3, 8  ...  17.0  PDFFUNC001   
12   2000-12-01 07:56:07              2, 4, 6  ...  19.0  MDGLIKE006   
14   2000-10-27 05:31:32                 5, 9  ...  18.0  MDGLIKE012   
...                  ...                  ...  ...   ...         ...   
4447 2001-07-05 02:30:13                    1  ...  22.0  MDEETCC015   
4451 2001-07-05 03:02:21        1, 3, 6, 7, 8  ...  21.0  PDFFUNC007   
4453 2001-03-06 11:42:36                 3, 6  ...  17.0  MDEHRWE008   
4455 2001-01-27 10:49:59     1, 2, 3, 4, 6, 7  ...  22.0  MDEHRWE008   
4456 2001-05-01 11:36:13           1, 2, 4, 8  ...  34.0  MDFMEAL007   

             p_name p_price           p_created  p_quantity age_group  \
6     파워세라믹종아리 다이어트   39000 2001-05-28 08:26:02           1        30   
7      뷰티라인(땡김이)-멀티   22000 2001-05-28 10:12:39           1        30   
11      내추럴하우스 라인-업   60000 2001-05-28 09:10:42           1        30   
12            매직마스크   24900 2001-05-27 01:36:40           1        30   
14     뷰티라인(땡김이)-심플   18000 2001-05-29 01:28:42           1        30   
...             ...     ...                 ...         ...       ...   
4447         스폐셜킥보드   35900 2001-07-05 02:36:53           1        30   
4451    깜짝변신 다이어트세트  130000 2001-07-05 03:09:10           1        40   
4453  파워세라믹종아리 다이어트   29900 2001-07-05 07:17:10           1        20   
4455  파워세라믹종아리 다이어트   29900 2001-07-06 01:17:45           1        20   
4456     대상다이어트프로그램   70000 2001-07-06 01:50:12           1        30   

      product buy season join season label  
6                      1           2     1  
7                      1           3     1  
11                     1           2     1  
12                     1           4     1  
14                     1           3     1  
...                  ...         ...   ...  
4447                   2           2     0  
4451                   2           2     0  
4453                   2           1     0  
4455                   2           4     0  
4456                   2           1     0  

[4458 rows x 22 columns]

In [20]:
#문자열 데이터를 수치화 할 수 없기에 ,를 삭제

df['diet'] = df['diet'].str.replace(",","")
df

id  social_id1  occupation  sex  height  weight  married dise  \
6         bluezoo    761220.0        17.0  2.0   167.0    58.0      0.0    2   
7     sunsook1120    751120.0        17.0  2.0   168.0    62.0      0.0    3   
11        minimee    740531.0        15.0  2.0   160.0    46.0      0.0    0   
12        hammy99    800808.0        15.0  2.0   164.0    51.0      0.0    0   
14           ippn    770415.0        17.0  2.0   160.0    47.0      0.0    2   
...           ...         ...         ...  ...     ...     ...      ...  ...   
4447         ekoo    770830.0        15.0  1.0   178.0    70.0      0.0    0   
4451      hjlee90    711023.0         7.0  2.0   172.0    64.0      1.0    0   
4453   0162655575    830219.0        15.0  2.0   165.0    49.0      0.0    0   
4455   onlyonelhm    880330.0        15.0  2.0   159.0    58.0      0.0    0   
4456    morebest1    760709.0        12.0  1.0   177.0   109.0      0.0    0   

                 created            loss_part  ...   bmi        pfid  \
6    2000-08-18 01:59:10           4, 5, 6, 8  ...  21.0  MDEPART007   
7    2000-10-20 11:32:54  1, 2, 3, 4, 5, 6, 7  ...  21.0  MDGLIKE013   
11   2000-06-14 11:16:57              1, 3, 8  ...  17.0  PDFFUNC001   
12   2000-12-01 07:56:07              2, 4, 6  ...  19.0  MDGLIKE006   
14   2000-10-27 05:31:32                 5, 9  ...  18.0  MDGLIKE012   
...                  ...                  ...  ...   ...         ...   
4447 2001-07-05 02:30:13                    1  ...  22.0  MDEETCC015   
4451 2001-07-05 03:02:21        1, 3, 6, 7, 8  ...  21.0  PDFFUNC007   
4453 2001-03-06 11:42:36                 3, 6  ...  17.0  MDEHRWE008   
4455 2001-01-27 10:49:59     1, 2, 3, 4, 6, 7  ...  22.0  MDEHRWE008   
4456 2001-05-01 11:36:13           1, 2, 4, 8  ...  34.0  MDFMEAL007   

             p_name p_price           p_created  p_quantity age_group  \
6     파워세라믹종아리 다이어트   39000 2001-05-28 08:26:02           1        30   
7      뷰티라인(땡김이)-멀티   22000 2001-05-28 10:12:39           1        30   
11      내추럴하우스 라인-업   60000 2001-05-28 09:10:42           1        30   
12            매직마스크   24900 2001-05-27 01:36:40           1        30   
14     뷰티라인(땡김이)-심플   18000 2001-05-29 01:28:42           1        30   
...             ...     ...                 ...         ...       ...   
4447         스폐셜킥보드   35900 2001-07-05 02:36:53           1        30   
4451    깜짝변신 다이어트세트  130000 2001-07-05 03:09:10           1        40   
4453  파워세라믹종아리 다이어트   29900 2001-07-05 07:17:10           1        20   
4455  파워세라믹종아리 다이어트   29900 2001-07-06 01:17:45           1        20   
4456     대상다이어트프로그램   70000 2001-07-06 01:50:12           1        30   

      product buy season join season label  
6                      1           2     1  
7                      1           3     1  
11                     1           2     1  
12                     1           4     1  
14                     1           3     1  
...                  ...         ...   ...  
4447                   2           2     0  
4451                   2           2     0  
4453                   2           1     0  
4455                   2           4     0  
4456                   2           1     0  

[4458 rows x 22 columns]

In [21]:
# diet 3,4,5,6 이 포함되어 있으면 1 없으면 0으로 표기

df['diet'] = df['diet'].map(lambda x : str(x))
diet_experience = [0 for i in df['diet']]
for i, diet in enumerate(df['diet']):
  if ('3' in diet) | ('4' in diet) | ('5' in diet)| ('6' in diet):
      #print(diet)
      diet_experience[i] = 1
df['diet_experience'] = diet_experience

In [22]:
#loss_part데이터를 1~9까지 새로운 columns을 하나씩 만듬

df["loss_part_1"] = df["loss_part"].str.contains("1", na = False)*1
df["loss_part_2"] = df["loss_part"].str.contains("2", na = False)*1
df["loss_part_3"] = df["loss_part"].str.contains("3", na = False)*1
df["loss_part_4"] = df["loss_part"].str.contains("4", na = False)*1
df["loss_part_5"] = df["loss_part"].str.contains("5", na = False)*1
df["loss_part_6"] = df["loss_part"].str.contains("6", na = False)*1
df["loss_part_7"] = df["loss_part"].str.contains("7", na = False)*1
df["loss_part_8"] = df["loss_part"].str.contains("8", na = False)*1
df["loss_part_9"] = df["loss_part"].str.contains("9", na = False)*1

In [23]:
df

id  social_id1  occupation  sex  height  weight  married dise  \
6         bluezoo    761220.0        17.0  2.0   167.0    58.0      0.0    2   
7     sunsook1120    751120.0        17.0  2.0   168.0    62.0      0.0    3   
11        minimee    740531.0        15.0  2.0   160.0    46.0      0.0    0   
12        hammy99    800808.0        15.0  2.0   164.0    51.0      0.0    0   
14           ippn    770415.0        17.0  2.0   160.0    47.0      0.0    2   
...           ...         ...         ...  ...     ...     ...      ...  ...   
4447         ekoo    770830.0        15.0  1.0   178.0    70.0      0.0    0   
4451      hjlee90    711023.0         7.0  2.0   172.0    64.0      1.0    0   
4453   0162655575    830219.0        15.0  2.0   165.0    49.0      0.0    0   
4455   onlyonelhm    880330.0        15.0  2.0   159.0    58.0      0.0    0   
4456    morebest1    760709.0        12.0  1.0   177.0   109.0      0.0    0   

                 created            loss_part  ... diet_experience  \
6    2000-08-18 01:59:10           4, 5, 6, 8  ...               1   
7    2000-10-20 11:32:54  1, 2, 3, 4, 5, 6, 7  ...               1   
11   2000-06-14 11:16:57              1, 3, 8  ...               1   
12   2000-12-01 07:56:07              2, 4, 6  ...               1   
14   2000-10-27 05:31:32                 5, 9  ...               1   
...                  ...                  ...  ...             ...   
4447 2001-07-05 02:30:13                    1  ...               0   
4451 2001-07-05 03:02:21        1, 3, 6, 7, 8  ...               0   
4453 2001-03-06 11:42:36                 3, 6  ...               0   
4455 2001-01-27 10:49:59     1, 2, 3, 4, 6, 7  ...               0   
4456 2001-05-01 11:36:13           1, 2, 4, 8  ...               1   

     loss_part_1  loss_part_2 loss_part_3 loss_part_4  loss_part_5  \
6              0            0           0           1            1   
7              1            1           1           1            1   
11             1            0           1           0            0   
12             0            1           0           1            0   
14             0            0           0           0            1   
...          ...          ...         ...         ...          ...   
4447           1            0           0           0            0   
4451           1            0           1           0            0   
4453           0            0           1           0            0   
4455           1            1           1           1            0   
4456           1            1           0           1            0   

     loss_part_6  loss_part_7 loss_part_8 loss_part_9  
6              1            0           1           0  
7              1            1           0           0  
11             0            0           1           0  
12             1            0           0           0  
14             0            0           0           1  
...          ...          ...         ...         ...  
4447           0            0           0           0  
4451           1            1           1           0  
4453           1            0           0           0  
4455           1            1           0           0  
4456           0            0           1           0  

[4458 rows x 32 columns]

In [24]:
df["height"].fillna(df["height"].mean(),inplace = True)
df["weight"].fillna(df["weight"].mean(),inplace = True)
df["bmi"].fillna(df["bmi"].mean(),inplace = True)

In [25]:
df.dropna(axis=0, inplace=True)

In [26]:
df.isnull().sum()

id                    0
social_id1            0
occupation            0
sex                   0
height                0
weight                0
married               0
dise                  0
created               0
loss_part             0
diet_purpose          0
diet                  0
bmi                   0
pfid                  0
p_name                0
p_price               0
p_created             0
p_quantity            0
age_group             0
product buy season    0
join season           0
label                 0
diet_experience       0
loss_part_1           0
loss_part_2           0
loss_part_3           0
loss_part_4           0
loss_part_5           0
loss_part_6           0
loss_part_7           0
loss_part_8           0
loss_part_9           0
dtype: int64

#전처리 끝

의사결정나무 - 뷰쳐임포턴스 뽑기 위해

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4455 entries, 6 to 4456
Data columns (total 32 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   id                  4455 non-null   object        
 1   social_id1          4455 non-null   float64       
 2   occupation          4455 non-null   float64       
 3   sex                 4455 non-null   float64       
 4   height              4455 non-null   float64       
 5   weight              4455 non-null   float64       
 6   married             4455 non-null   float64       
 7   dise                4455 non-null   object        
 8   created             4455 non-null   datetime64[ns]
 9   loss_part           4455 non-null   object        
 10  diet_purpose        4455 non-null   object        
 11  diet                4455 non-null   object        
 12  bmi                 4455 non-null   float64       
 13  pfid                4455 non-null   object      

In [28]:
df["age_group"] = df["age_group"].astype(int)
df["product buy season"] = df["product buy season"].astype(int)
df["join season"] = df["join season"].astype(int)

In [29]:
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
from sklearn.metrics import classification_report

X = df[['occupation','diet_purpose', 'bmi', 'p_price', 'p_quantity','product buy season',
       'diet_experience',"join season",'loss_part_1', 'loss_part_2', 'loss_part_3', 'loss_part_4',
       'loss_part_5', 'loss_part_6', 'loss_part_7', 'loss_part_8','loss_part_9']]
Y = df['label']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=0)

smote = SMOTE(random_state = 0)
X_train, Y_train = smote.fit_resample(X_train, Y_train)

from sklearn.tree import DecisionTreeClassifier

DT_model = DecisionTreeClassifier(random_state = 0, max_depth = 3)
DT_model.fit(X_train, Y_train)
Y_pred = DT_model.predict(X_test)
print("accuracy : {:.3f}".format(DT_model.score(X_train, Y_train)))
print(classification_report(Y_test, Y_pred))

accuracy : 0.638
              precision    recall  f1-score   support

           0       0.52      0.72      0.60       564
           1       0.72      0.51      0.60       773

    accuracy                           0.60      1337
   macro avg       0.62      0.62      0.60      1337
weighted avg       0.63      0.60      0.60      1337



In [30]:
feature_name = X.columns
feature_importance = DT_model.feature_importances_
sorted(zip(feature_importance,feature_name), reverse = True)

[(0.3519980036714569, 'join season'),
 (0.24925091188296528, 'diet_purpose'),
 (0.2437622270169359, 'occupation'),
 (0.13746414114233194, 'loss_part_6'),
 (0.01752471628630998, 'p_price'),
 (0.0, 'product buy season'),
 (0.0, 'p_quantity'),
 (0.0, 'loss_part_9'),
 (0.0, 'loss_part_8'),
 (0.0, 'loss_part_7'),
 (0.0, 'loss_part_5'),
 (0.0, 'loss_part_4'),
 (0.0, 'loss_part_3'),
 (0.0, 'loss_part_2'),
 (0.0, 'loss_part_1'),
 (0.0, 'diet_experience'),
 (0.0, 'bmi')]

#SVM 모델

In [31]:
X = df[['occupation','diet_purpose', 'bmi', 'p_price', 'p_quantity','product buy season',
       'diet_experience',"join season",'loss_part_1', 'loss_part_2', 'loss_part_3', 'loss_part_4',
       'loss_part_5', 'loss_part_6', 'loss_part_7', 'loss_part_8','loss_part_9']]
Y = df['label']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=0)

smote = SMOTE(random_state = 0)
X_train, Y_train = smote.fit_resample(X_train, Y_train)

from sklearn.svm import SVC

SVC_model = SVC(C=1000, gamma = 10, random_state = 0)  #C 값은 클수록 이상치를 포함하는 것이고 적을 수록 이상치를 버리는 것
SVC_model.fit(X_train, Y_train)                        #gamma는 첨도와 같은 것 데이터에 가깝게 표현해서 원만한 선이 뾰족해진다
Y_pred = SVC_model.predict(X_test)
print("accuracy : {:.3f}".format(SVC_model.score(X_train, Y_train)))
print(classification_report(Y_test, Y_pred))

accuracy : 0.994
              precision    recall  f1-score   support

           0       0.92      0.20      0.33       564
           1       0.63      0.99      0.77       773

    accuracy                           0.65      1337
   macro avg       0.77      0.59      0.55      1337
weighted avg       0.75      0.65      0.58      1337



#로지스틱 회귀분석

In [32]:
from sklearn.preprocessing import StandardScaler

X = df[['occupation','diet_purpose', 'bmi', 'p_price', 'p_quantity','product buy season',
       'diet_experience',"join season",'loss_part_1', 'loss_part_2', 'loss_part_3', 'loss_part_4',
       'loss_part_5', 'loss_part_6', 'loss_part_7', 'loss_part_8','loss_part_9']]
Y = df['label']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=0)

scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)
print(X_train)

#oversampling
smote = SMOTE(random_state=0)
X_train_over, Y_train_over = smote.fit_resample(X_train, Y_train)

[[-0.05007161 -0.74251742 -0.35974719 ... -0.94263057 -0.38855916
  -0.442731  ]
 [ 0.5605801  -0.74251742 -0.1572392  ... -0.94263057 -0.38855916
  -0.442731  ]
 [ 0.5605801  -0.74251742  0.24777677 ... -0.94263057 -0.38855916
  -0.442731  ]
 ...
 [ 0.5605801  -0.74251742 -0.56225518 ...  1.06086099 -0.38855916
  -0.442731  ]
 [ 0.35702953 -0.74251742  0.04526878 ... -0.94263057 -0.38855916
  -0.442731  ]
 [ 0.5605801  -0.74251742  0.24777677 ... -0.94263057 -0.38855916
  -0.442731  ]]


In [33]:
from sklearn.linear_model import LogisticRegression

logit = LogisticRegression(C = 1, random_state = 0)

logit.fit(X_train_over, Y_train_over)
Y_pred = logit.predict(X_test)

#Y_pred
print("R2 : {:.3f}".format(logit.score(X_train_over, Y_train_over)))

from sklearn.metrics import classification_report
print(classification_report(Y_test, Y_pred))
print(logit.intercept_)
print(logit.coef_)

R2 : 0.632
              precision    recall  f1-score   support

           0       0.53      0.63      0.58       564
           1       0.69      0.60      0.64       773

    accuracy                           0.61      1337
   macro avg       0.61      0.62      0.61      1337
weighted avg       0.62      0.61      0.61      1337

[0.02958165]
[[-0.22034002 -0.2269007  -0.1151369   0.01120754 -0.21594382 -0.12605003
   0.19924666  0.30442819 -0.02625905 -0.21883164 -0.06202047  0.02081032
   0.12280575  0.24666674  0.18720365  0.01146252 -0.07983759]]


In [34]:
from sklearn.metrics import accuracy_score

# 로지스틱 회귀 모델의 예측값을 계산합니다.
Y_pred = logit.predict(X_test)

# Accuracy를 계산합니다.
accuracy = accuracy_score(Y_test, Y_pred)

# Accuracy를 출력합니다.
print('Accuracy: {:.3f}'.format(accuracy))

Accuracy: 0.613


In [35]:
df['occupation'] = df['occupation'].astype(int)
df['bmi'] = df['bmi'].astype(int)
df['diet_purpose']= df['diet_purpose'].astype(int)

#랜덤 포레스트

In [36]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder

#1.변수선택
X = df[['occupation','diet_purpose', 'bmi', 'p_price', 'p_quantity','product buy season',
       'diet_experience',"join season",'loss_part_1', 'loss_part_2', 'loss_part_3', 'loss_part_4',
       'loss_part_5', 'loss_part_6', 'loss_part_7', 'loss_part_8','loss_part_9']]
Y = df['label']
#2.train - test split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=0)

#3.데이터 전처리(preprocessing)
ct = ColumnTransformer([("scaling", StandardScaler(), ['bmi','p_price', 'p_quantity']),
                        ("onehot", OneHotEncoder(sparse = False), ['occupation','diet_purpose','diet_experience',"join season",'product buy season','loss_part_1', 'loss_part_2', 'loss_part_3', 'loss_part_4',
       'loss_part_5', 'loss_part_6', 'loss_part_7', 'loss_part_8','loss_part_9'])])
ct.fit(X)
X_train = ct.transform(X_train)
X_test = ct.transform(X_test)

#4. 오버샘플링
smote = SMOTE(random_state=0)
X_train, Y_train = smote.fit_resample(X_train, Y_train)

/usr/local/lib/python3.9/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [37]:
#5. 모형 생성
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier(random_state = 0, n_estimators= 300, max_depth = 3)
model. fit(X_train,Y_train)
Y_pred = model.predict(X_test)
print("Y_predicti value: \n", Y_pred)
print("accuracy(test) : {:.3f}".format(model.score(X_test, Y_test)))
print(classification_report(Y_test, Y_pred))

Y_predicti value: 
 [1 1 1 ... 1 1 1]
accuracy(test) : 0.707
              precision    recall  f1-score   support

           0       0.69      0.55      0.61       564
           1       0.72      0.82      0.76       773

    accuracy                           0.71      1337
   macro avg       0.70      0.69      0.69      1337
weighted avg       0.70      0.71      0.70      1337



In [38]:
feature_name = X.columns
feature_importance = model.feature_importances_
sorted(zip(feature_importance,feature_name), reverse = True)

[(0.30925653228176897, 'loss_part_9'),
 (0.043344402925499154, 'occupation'),
 (0.040268560077217214, 'product buy season'),
 (0.01577633470041623, 'diet_purpose'),
 (0.008973120745337003, 'loss_part_7'),
 (0.003877422545537981, 'loss_part_1'),
 (0.0022015956161699412, 'bmi'),
 (0.0011293497436949502, 'loss_part_2'),
 (0.0005091061699990983, 'p_price'),
 (0.00044630283288591246, 'loss_part_3'),
 (0.0002846317886640759, 'loss_part_8'),
 (0.0002835785267881448, 'diet_experience'),
 (0.0, 'p_quantity'),
 (0.0, 'loss_part_6'),
 (0.0, 'loss_part_5'),
 (0.0, 'loss_part_4'),
 (0.0, 'join season')]

#KNN

In [39]:
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder


X = df[['occupation', 'diet_purpose', 'bmi', 'p_price', 'p_quantity','product buy season',
       'diet_experience',"join season",'loss_part_1', 'loss_part_2', 'loss_part_3', 'loss_part_4',
       'loss_part_5', 'loss_part_6', 'loss_part_7', 'loss_part_8',
       'loss_part_9']]
Y = df['label']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3, random_state = 0)

scaler = StandardScaler().fit(X_train, Y_train)
X_test = scaler.transform(X_test)
X_train = scaler.transform(X_train)

X_train, Y_train = SMOTE(random_state = 0).fit_resample(X_train, Y_train)

from sklearn.neighbors import KNeighborsClassifier
knn_model = KNeighborsClassifier(n_neighbors = 9, metric = "euclidean") #euclidean - 두점사이 거리 공식

knn_model.fit(X_train, Y_train)
Y_pred = knn_model.predict(X_test)
print("accuracy : {:.3f}".format(knn_model.score(X_train, Y_train)))
print(classification_report(Y_test, Y_pred)) #accuracy, f1 score 보는게 가장 정확
#실제 f1 score : 아래의 0과 1 f1 score의 평균값
#맨위 accuracy - train 값 (train 데이터 잘 설명 하는지)
#아래 accuracy - test pred 계산했을때 나온 값 (예측잘 하는지)

accuracy : 0.754
              precision    recall  f1-score   support

           0       0.54      0.64      0.59       564
           1       0.70      0.61      0.65       773

    accuracy                           0.62      1337
   macro avg       0.62      0.62      0.62      1337
weighted avg       0.63      0.62      0.62      1337



#나이브베이즈

In [40]:
X = df[['occupation', 'diet_purpose', 'bmi', 'p_price', 'p_quantity','product buy season',
       'diet_experience',"join season",'loss_part_1', 'loss_part_2', 'loss_part_3', 'loss_part_4',
       'loss_part_5', 'loss_part_6', 'loss_part_7', 'loss_part_8',
       'loss_part_9']]
Y = df['label']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=0)

smote = SMOTE(random_state = 0)
X_train, Y_train = smote.fit_resample(X_train, Y_train)

from sklearn.naive_bayes import GaussianNB

NB_model = GaussianNB(var_smoothing = 0)  
NB_model.fit(X_train, Y_train)                     
Y_pred = NB_model.predict(X_test)
print("accuracy : {:.3f}".format(NB_model.score(X_train, Y_train)))
print(classification_report(Y_test, Y_pred))

accuracy : 0.547
              precision    recall  f1-score   support

           0       0.59      0.15      0.23       564
           1       0.60      0.92      0.73       773

    accuracy                           0.60      1337
   macro avg       0.59      0.54      0.48      1337
weighted avg       0.59      0.60      0.52      1337



#인공신경망


In [41]:
#1.변수선택
X = df[['occupation','diet_purpose', 'bmi', 'p_price', 'p_quantity','product buy season',
       'diet_experience',"join season",'loss_part_1', 'loss_part_2', 'loss_part_3', 'loss_part_4',
       'loss_part_5', 'loss_part_6', 'loss_part_7', 'loss_part_8','loss_part_9']]
Y = df['label']
#2.train - test split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=0)

#3.데이터 전처리(preprocessing)
ct = ColumnTransformer([("scaling", StandardScaler(), ['bmi','p_price', 'p_quantity']),
                        ("onehot", OneHotEncoder(sparse = False), ['occupation','diet_purpose','diet_experience',"join season",'product buy season','loss_part_1', 'loss_part_2', 'loss_part_3', 'loss_part_4',
       'loss_part_5', 'loss_part_6', 'loss_part_7', 'loss_part_8','loss_part_9'])])
ct.fit(X)
X_train = ct.transform(X_train)
X_test = ct.transform(X_test)

#4. 오버샘플링
smote = SMOTE(random_state = 0)
X_train, Y_train = smote.fit_resample(X_train, Y_train)

/usr/local/lib/python3.9/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [42]:
#5. 모델링
from sklearn.neural_network import MLPClassifier
nn_model = MLPClassifier(random_state =0, alpha = 0.001, hidden_layer_sizes = [50]) #layer 만들었지만 순차적x 주르륵 늘어뜨림
##5-1. 모형학습
nn_model.fit(X_train, Y_train)

#6. 예측
Y_pred = nn_model.predict(X_test)


/usr/local/lib/python3.9/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


In [43]:
#7.결과값 보고
print("Y 예측값 \n", Y_pred)
print("accuracy(train) : {:.3f}".format(nn_model.score(X_train, Y_train)))
from sklearn.metrics import classification_report
print(classification_report(Y_test, Y_pred))

Y 예측값 
 [1 1 1 ... 0 0 1]
accuracy(train) : 0.840
              precision    recall  f1-score   support

           0       0.66      0.70      0.68       564
           1       0.77      0.74      0.75       773

    accuracy                           0.72      1337
   macro avg       0.71      0.72      0.72      1337
weighted avg       0.72      0.72      0.72      1337



#그래디언 부스트

In [74]:
#그레디언트 부스팅

#1.변수선택
X = df[['occupation','diet_purpose', 'bmi', 'p_price','product buy season',
       'diet_experience','loss_part_1', 'loss_part_2', 'loss_part_3', 'loss_part_4',
       'loss_part_5', 'loss_part_6', 'loss_part_7', 'loss_part_8','loss_part_9']]
Y = df['label']
#2.train - test split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=0)

#3.데이터 전처리(preprocessing)
ct = ColumnTransformer([("scaling", StandardScaler(), ['bmi','p_price']),
                        ("onehot", OneHotEncoder(sparse = False), ['occupation','diet_purpose','diet_experience','product buy season','loss_part_1', 'loss_part_2', 'loss_part_3', 'loss_part_4',
       'loss_part_5', 'loss_part_6', 'loss_part_7', 'loss_part_8','loss_part_9'])])
ct.fit(X)
X_train = ct.transform(X_train)
X_test = ct.transform(X_test)

#4. 오버샘플링
smote = SMOTE(random_state = 0)
X_train, Y_train = smote.fit_resample(X_train, Y_train)

/usr/local/lib/python3.9/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [75]:
#5. 모형 생성
from sklearn.ensemble import GradientBoostingClassifier
model = GradientBoostingClassifier(random_state = 0, n_estimators = 100, max_depth = 4, learning_rate = 0.1)
#learning_rate = 전 모델을 얼마나 반영할지의 수치

#6. 모형 학습 예측
model.fit(X_train, Y_train)
Y_pred = model.predict(X_test)
print("Y predict value : \n" , Y_pred)
print("accuracy(test) : {:.3f}".format(model.score(X_test, Y_test)))
print(classification_report(Y_test, Y_pred))

Y predict value : 
 [1 1 1 ... 1 0 1]
accuracy(test) : 0.716
              precision    recall  f1-score   support

           0       0.69      0.60      0.64       564
           1       0.73      0.80      0.77       773

    accuracy                           0.72      1337
   macro avg       0.71      0.70      0.70      1337
weighted avg       0.71      0.72      0.71      1337



In [76]:
feature_name = X.columns
feature_importance = model.feature_importances_
sorted(zip(feature_importance,feature_name), reverse = True)

[(0.12590662697569371, 'diet_purpose'),
 (0.09211442896404297, 'occupation'),
 (0.014314889283410796, 'product buy season'),
 (0.00957149407247584, 'loss_part_2'),
 (0.008856139685865462, 'loss_part_8'),
 (0.00691542216369073, 'bmi'),
 (0.005633266732703051, 'loss_part_3'),
 (0.0049453779076393135, 'diet_experience'),
 (0.0025740208358714044, 'loss_part_9'),
 (0.0022759572968588525, 'loss_part_4'),
 (0.0011842931494038893, 'loss_part_6'),
 (0.001125578145329096, 'loss_part_7'),
 (0.0006168716950892173, 'loss_part_5'),
 (0.0, 'p_price'),
 (0.0, 'loss_part_1')]